File: TeoriaDeAlgoritmos/src/main/java/ar/fi/uba/tda/colecciones/Vertice.java
package ar.fi.uba.tda.colecciones;

public class Vertice<T> {

	private T contenido; 
	private boolean visitado;
	private ListaEnlazada<Vertice<T>> adyacentes;
	private Long index;
	private Long lowLink;
	
		
	public Vertice() {
		super();
	}
	
	public Vertice(T contenido){
		super();
		this.contenido=contenido;
		this.adyacentes=new ListaEnlazada<Vertice<T>>();
	}
	
	public T getContenido() {
		return contenido;
	}
	public void setContenido(T contenido) {
		this.contenido = contenido;
	}
	public ListaEnlazada<Vertice<T>> getAdyacentes() {
		return adyacentes;
	}
	public void setAdyacentes(ListaEnlazada<Vertice<T>> adyacentes) {
		this.adyacentes = adyacentes;
	}

	public Long getIndex() {
		return index;
	}

	public void setIndex(Long index) {
		this.index = index;
	}

	public Long getLowLink() {
		return lowLink;
	}

	public void setLowLink(Long lowLink) {
		this.lowLink = lowLink;
	}

	public boolean isVisitado() {
		return visitado;
	}

	public void setVisitado(boolean vistado) {
		this.visitado = vistado;
	}
	
	public Integer getGradoVertice(){
		return this.adyacentes.tamanio();
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result
				+ ((contenido == null) ? 0 : contenido.hashCode());
		return result;
	}

	@Override
	@SuppressWarnings("unchecked")
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (!(obj instanceof Vertice)) {
			return false;
		}
		Vertice<T> other = (Vertice<T>) obj;
		if (contenido == null) {
			if (other.contenido != null) {
				return false;
			}
		} else if (!contenido.equals(other.contenido)) {
			return false;
		}
		return true;
	}

	@Override
	public String toString() {
		return "Vertice: "+contenido.toString()+" (i: "+this.getIndex()+" , l: "+this.getLowLink()+")";
	}
	
}

File: Robustez/Robustez/Vertice.cs
﻿using System;

namespace Robustez
{
    public enum Color {blanco, gris, negro};

    public class Vertice<T>
    {
        private T _contenido;
        private bool _visitado;
        private ListaEnlazada<Vertice<T>> _adyacentes;
        private Color _color;
        private Vertice<T> _padre;
        private long _index;
        private long _lowLink;
        private bool _agregadoEnListaCiclo;

        public bool AgregadoEnListaCiclo
        {
            get { return _agregadoEnListaCiclo; }
            set { _agregadoEnListaCiclo = value; }
        }

        //public bool AgregadoEnListaCiclo { get; set; }

        public long Index
        {
            get { return _index; }
            set { _index = value; }
        }

        public long LowLink
        {
            get { return _lowLink; }
            set { _lowLink = value; }
        }

        public Color Color
        {
            get { return _color; }
            set { _color = value; }
        }

        public Vertice<T> Padre
        {
            get { return _padre; }
            set { _padre = value; }
        }

        public T Contenido
        {
            get { return _contenido; }
            set { _contenido = value; }
        }


        public bool Visitado
        {
            get { return _visitado; }
            set { _visitado = value; }
        }


        public ListaEnlazada<Vertice<T>> Adyacentes
        {
            get { return _adyacentes; }
            set { _adyacentes = value; }
        }

        public Vertice()
        {
            Adyacentes = new ListaEnlazada<Vertice<T>>();
        }

        public Vertice(T contenido)
        {
            Contenido = contenido;
            Adyacentes = new ListaEnlazada<Vertice<T>>();
        }

        
        public Int32 GetGradoVertice()
        {
            return _adyacentes.Tamanio;
        }



        public override int GetHashCode()
        {
            const int prime = 31;
            int result = 1;
            result = prime * result
                    + ((Contenido == null) ? 0 : Contenido.GetHashCode());
            return result;
        }

        /// <summary>
        /// Devuelve true o false segun si un vertice es igual a otro respectivamente.
        /// </summary>
        /// <param name="obj"></param>
        /// <returns></returns>
        public override bool Equals(Object obj)
        {
            if (this == null && obj == null)
                return true;
            else
            {
                if (this == null || obj == null)
                    return false;
                else
                {
                    Vertice<T> otroVertice = (Vertice<T>) obj;
                    if (Contenido == null && otroVertice.Contenido == null)
                        return true;
                    else
                    {
                        if (Contenido == null || otroVertice.Contenido == null)
                            return false;
                        else
                            return Contenido.Equals(otroVertice.Contenido);
                    }

                }
            }
        }
        


        /// <summary>
        /// Metodo sobreescrito para mostrar claramente el vertice.
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            return "Vertice: " + Contenido + " (i: " + Index + " , l: " + LowLink + ")";
        }

    }
}
